orttraining/orttraining/python/training/_ortmodule_execution_manager.py

```diff
@@ -0,0 +1,268 @@
+                                  do_constant_folding=False,
+                                  training=training_flag,
+                                  dynamic_axes=dynamic_axes,
+                                  verbose=self._verbosity < _ortmodule_utils.Verbosity.WARNING,
+                                  export_params=False,
+                                  keep_initializers_as_inputs=True)
+        except RuntimeError as e:
+            raise RuntimeError(
+                'There was an error while exporting the PyTorch model to ONNX: {}'.format(e))
+
+        return onnx.load_model_from_string(f.getvalue())
+
+    def _set_device_from_module(self):
+        """Get the device from the module and save it to self._device"""
+
+        device_from_module = _utils.get_device_from_module(
+            self._original_module)
+        if not self._device or self._device != device_from_module:
+            self._device = device_from_module
+            if not self._device:
+                raise RuntimeError(
+                    'A device must be specified in the model or data!')
+
+    def _convert_training_graph_input_to_list(self, *inputs, **kwargs):
+        '''Creates forward `*inputs` list from user input and PyTorch initializers
+
+        TODO: How IO binding model inputs and outputs affects initializer copies?
+
+        ONNX Runtime forward requires an ordered list of:
+            * User input: computed from forward InferenceSession
+            * Initializers: computed from original PyTorch model parameters
+        '''
+        # User inputs
+        non_none_inputs = [inp for inp in inputs if inp is not None]
+        named_buffers_iter = iter(self._flattened_output_module.named_buffers())
+        result = []
+        for input_idx, name in enumerate(self._graph_info.user_input_names):
+            inp = None
+            if input_idx < len(non_none_inputs):
+                inp = non_none_inputs[input_idx]
+            elif name in kwargs and kwargs[name] is not None:
+                inp = kwargs[name]
+            elif input_idx >= len(non_none_inputs):
+                # Registered buffers are translated to user_input+initializer in ONNX
+                # TODO: Check what happens when the number of inputs change form one call to the next
+                buffer_name, inp = next(named_buffers_iter)
+                assert buffer_name == name, f'Input name {name} expected, but {buffer_name} found!'
+
+            if inp is not None:
+                result.append(inp)
+            else:
+                # TODO: Re-export ONNX if any input from onnx_graphs_info.user_input_names is None.
+                raise RuntimeError(
+                    f'Input is present in ONNX graph but not provided: {name}.')
+
+        # Initializers
+        for param in self._flattened_output_module.named_parameters():
+            result.append(param[1])
+
+        return result
+
+    def _initialize_module_graph_builder(self):
```

Comment:
Can this be deferred to `TrainingManager`? It would remove the need for the is_training check in the base class

Reply:
I can do that. The function pretty much is the same between the two Managers, except for the training flag being passed to the backend. I think I can call the base class method with the training flag as an argument to avoid needing _is_training check in the base class.

Reply:
Yeah, I was guessing we couldnt get rid of this without redundancy because of a single flag.
I am not 100% convinced we need to move this to the specialized classes